mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,181 +166,73 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
             auto it = std::find(diffArgIndices.begin(), diffArgIndices.end(), index);
             if (it == diffArgIndices.end()) {
-                auto const_global = getOrInsertEnzymeConstDecl(rewriter, op);
-                auto llvmI8PtrTy =
-                    LLVM::LLVMPointerType::get(IntegerType::get(op->getContext(), 8));
-                auto enzymeConst =
-                    rewriter.create<LLVM::AddressOfOp>(op->getLoc(), llvmI8PtrTy, const_global);
-                callArgs.push_back(enzymeConst);
-                // Add the argument to call arg as a pointer
-                auto newMemrefArg = rewriter.create<LLVM::AllocaOp>(
-                    loc, LLVM::LLVMPointerType::get(llvmMemrefArg.getType()), c1);
-                rewriter.create<LLVM::StoreOp>(loc, llvmMemrefArg, newMemrefArg);
-                callArgs.push_back(newMemrefArg);
-            }
-            else {
-                auto position = std::distance(diffArgIndices.begin(), it);
-                auto llvmDataIn = dataIn[position];
-                // Get information about the memref arg
-                MemRefType memrefArgType = memrefArg.getType().cast<MemRefType>();
-                size_t sizeShape = memrefArgType.getShape().size();
-                size_t value = memrefArgType.getElementTypeBitWidth() / 8;
-                auto memrefArgSize =
-                    rewriter.create<LLVM::ConstantOp>(loc, IntegerType::get(ctx, 64), value);
-
-                // Add the argument to call arg as a pointer
-                auto newMemrefArg = rewriter.create<LLVM::AllocaOp>(
-                    loc, LLVM::LLVMPointerType::get(llvmMemrefArg.getType()), c1);
-                rewriter.create<LLVM::StoreOp>(loc, llvmMemrefArg, newMemrefArg);
-                callArgs.push_back(newMemrefArg);
-
-                // Shadow of args
-                Value bufferSize;
-
-                if (sizeShape == 0) {
-                    bufferSize =
-                        rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(1));
+                if (isa<MemRefType>(arg.getType())) {
+                    unpackMemRef(arg, /*shadow=*/nullptr, callArgs, rewriter, loc);
                 }
                 else {
-                    Value offset = rewriter.create<LLVM::ExtractValueOp>(loc, llvmMemrefArg, 2);
-
-                    SmallVector<int64_t> vectorIndices{3, 0};
-
-                    Value sizeArray =
-                        rewriter.create<LLVM::ExtractValueOp>(loc, llvmMemrefArg, vectorIndices);
-
-                    vectorIndices[0] = 4;
-                    Value strideArray =
-                        rewriter.create<LLVM::ExtractValueOp>(loc, llvmMemrefArg, vectorIndices);
-
-                    bufferSize = rewriter.create<LLVM::MulOp>(loc, sizeArray, strideArray);
-                    bufferSize = rewriter.create<LLVM::AddOp>(loc, offset, bufferSize);
+                    callArgs.push_back(castToConvertedType(arg, rewriter, loc));
```

Comment:
Maybe add a comment here to say that we do not need to make them constant because the type would be int or float. It is considered as non diff by Enzyme

Reply:
Actually, you raise a good point. If the argument is a float, Enzyme will assume it's an `enzyme_out` argument and compute an unnecessary derivative for it. I'll add an explicit `enzyme_const` annotation here.

Reply:
Oh cool, does it mean you were able to generate the Enzyme tags in the llvm dialect? 🤩  I remember it looked like that was not going to be possible.

Reply:
Yes, through an intermediate pass called `preserve-nvvm` in Enzyme that generates the required metadata from constructs accessible in the LLVM dialect.